[READY] [Gamemode Bounty] Assimilation

Kierany9 opened the pull request:
:cl: Kierany9
add: A new gamemode called Assimilation.
add: Several probably-inhuman creatures wielding psionic powers, known only as Hivemind Hosts, have appeared onboard the station and must assimilate as many crew members as possible. Assimilated crew serve as vessels for the Host to use their powers through (and on), and as a Host gains vessels, they gain new abilities.
remove: Vessels are not antagonists. Assimilation does not convert crew in any way.
add: Assimilated crewmembers can escape this nightmare by either getting a mindshield implant or by dying. Mindshielding vessels will reveal the identity of whoever assimilated them. Mindshielded crew are initially protected, but should a Host get strong enough, they can break these implants with ease.
add: Nothing is stopping Hosts from working together, but as their hives grow, so does the incentive to backstab. At 8 vessels, Hosts can probe the minds of crew and discover the identity of any other Hosts who have assimilated them, and at 18, they can steal every single vessel from a captured enemy Host.
tweak: Assimilation requires 25 players roundstart.
/:cl:

# PR is complete and ready to merge

## For the code-illiterate

- [x] Gamemode logic
- [x] Objectives
- [x] Cool powers
- [x] (Coder) Sprites
- [x] More cool powers that may or may not get added
- [x] Admin logging
- [x] Whatever the maintainers want done/fixed

### Another conversion mode?
No. Assimilated crew do not become antagonists, nor does it grant them any benefits. Assimilation only serves to fuel the power of and conflict between the roundstart antagonists.

### What objectives do these antagonists get?
- Assimilate X crewmembers OR Have X assimilated crewmembers escape on the shuttle
- Assimilate a crewmember(more likely than usual to be an implanted one) OR Kill/Maroon a crewmember
- Escape

### What powers do these "Hive Hosts" get?
All Hive Hosts begin with four powers, **Assimilate Vessel** allows them to add a crewmember within four tiles to the hive after standing still for ten seconds(and the crewmember needs to stay still for the last three). The power is undetectable, making it very safe to use. **Release Vessel** works the same way as Assimilate does, except it's instant and removes a hive member from the hive. **Hive Sight** allows them to use the eyes of an assimilated crew member, and **Neural Shock** knocks an assimilated target out for a short while, although the effectiveness of this power decreases with distance.

Later on, Hosts will unlock **Repair Protocol**, which allows them to absorb brain power from their vessels to heal brute, burn, brain and clone damage, ~~**Cerebellic Pulse** lets them move members around with effectiveness decreasing with distance~~, **Induce Panic** causes anybody in sight to hallucinate, drop everything or start shooting/whacking/hugging everybody in the room, **Network Invasion** lets them find out the identity of any other hive hosts connected to a crewmember, although this backfires if used against a host, **Mind Control** grants total control over a vessel for a short period of time, that grows rapidly with hive size and **Distortion Field** causes hallucinations (and later stamina damage) to everybody surrounding a vessel. 

Finally, **Bruteforce** destroys implants and nanites inside mindshielded personnel, **Telekinetic Field** is a stronger Forcewall, **Mass Assimilation** lets them steal an enemy host's vessels should you figure out their identity and capture them, and **Medullary Failure** causes a vessel to  ~~keel over and die~~ suffer a heart attack.

I've got a couple more in mind (that may or may not upset the entire balance of the game) but I'd like to see how this plays out as is first.

### Wow, that seems OP. What's stopping mass murderbone?
You lose your powers if you dip below the required hive size, the cooldowns are kind of long, many powers have a channelling time, and there are a grand total of two powers that affect non-assimilated personnel, so these powers are best used for other causes.

### What are you even trying to do with this mode?
A different take on conversion modes and stealthy antagonists with a large dose of espionage. Hosts have very little power against those they haven't assimilated, meaning the number one priority is to fill up the hive, both for the objective and for power in general, meaning they'll want to stay somewhat under the radar until unlocking at least Induce Panic. They also rely far more on information to survive than other antagonists, and are given the means to obtain it with powers such as Hive Sight and Network Invasion. Vessels are not antagonists, meaning that unlike standard conversion modes, there is no "Crew VS Enemies" dynamic that always crops up, in general crew will only end up involved in the mess if they happen to be a shared vessel between two hosts.

### Can I see your codersprites
![image](https://user-images.githubusercontent.com/11261264/45590765-c3462600-b940-11e8-94d4-c8ff0fa82431.png)
coder/10

## For maintainers
You may or may not want to tear your hair out reviewing this. I am sorry.

Kierany9 commented:
>accidentally grabbed a branch from 2016 and committed every single update since then after deleting it on remote

pr's off to a good start i see

as334 commented:
10 seconds within 4 tiles seems like it would be pretty difficult to pull off after the crew has sounded the alarm about the Hiveminds. Seems like it would be pretty obvious and just cause people to never stand still and be suspicious of anyone who does.

optimumtact commented:
Wow a shit remake of changeling good job.

naksu really is 2/3 for remaking existing gamemodes well done

Militaires commented:
oranges did u read past the title

Barhandar commented:
Mush when

zamolxius commented:
looks great and artfully made names,  its cool

fen0s commented:
> oranges did u read past the title

He can't really read. Can someone teach him maybe how to read letters? I have no idea how he's typing out words.

MrDoomBringer commented:
wouldnt the crewmembers be the "hosts" in this situation? the naming is weirding me out a bit

AnturK commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,462 @@
+			to_chat(user, "<span class='notice'>We begin linking our mind with [target.name]!</span>")
+			if(do_mob(user,user,70))
+				to_chat(user, "<span class='notice'>Our mind is ready to connect [target.name] to the Hive!</span>")
+				if((target in oview(range)) && do_mob(user,target,30))
```
why ignoring the user turf ?

AnturK commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,462 @@
+		power = max(5-(round(get_dist(user, target)/40)),2)
+		if(user.getCloneLoss()) //If we have genetic damage, prioritize this
+			user.adjustCloneLoss(-power)
+		else if(user.getBruteLoss() || user.getFireLoss()) //Otherwise heal the type we have more of
```
There's a helper for ordered heal that takes amount and damage type order.

AnturK commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,462 @@
+				charge_counter = charge_max
+		for(var/datum/antagonist/hivemind/enemy in GLOB.antagonists)
+			if(enemy.hivemembers.Find(target))
+				enemies += enemy.owner.name
```
You might want real names here, getting Unknown's sounds counterproductive.

Therapist1 commented:
Can not even my aluminum foil hat keep me safe?

ExcessiveUseOfCobblestone commented:
I haven't read the code but how do you find the bad guy as a not-bad guy? I mean once you randomly take brain damage you should immediately get mindshielded, but then what?

Kagus commented:
So, this is about silently and sneakily using the regular crewmembers as unwitting pawns? Sounds neat. Like a couple syndie comms agents facing off against each other, but with more Psychic Detective.

iskyp commented:
shadowling

Kierany9 commented:
Fixed everything in @AnturK's code review. Most spells already have sanity checks, I'll see if I've missed any later.
Replaced all `charge_counter = charge_max` with `revert_cast()`

>10 seconds within 4 tiles seems like it would be pretty difficult to pull off after the crew has sounded the alarm about the Hiveminds. Seems like it would be pretty obvious and just cause people to never stand still and be suspicious of anyone who does.

The alarm won't go off for quite a while, as they have zero tells unless they play horribly or get into an earlygame scuffle(although until you hit Network Invasion that's very unlikely), although we'll see if it ends up too weak in testing. Although the idea of a crew jittering about at all times is really funny.

>wouldnt the crewmembers be the "hosts" in this situation? the naming is weirding me out a bit

I was thinking of host in the 'network host' sense, not so much in the 'parasite-host dynamic' sense. I'm absolutely awful with names in general though, Assimilation was a last-minute idea anyway, might call them Assimilators if more people feel this way.

>I haven't read the code but how do you find the bad guy as a not-bad guy? I mean once you randomly take brain damage you should immediately get mindshielded, but then what?

Getting mindshielded will give you the host's name. I have no idea how this'll play out in testing, although I suspect it may be too strong and have a few nerfs in mind.

Kierany9 commented:
Had a first playtest, went pretty well, taking any feedback
* Fixed network invasion
  * No longer runtimes
  * Timers are now 150/300 deciseconds instead of 15/30
  * No longer detects your own hive as an enemy
* Added some info on other powers indicating that they require channelling or range
* Fixed objective requirement not being integers

Supermichael777 commented:
You need some incentive for a loyalty implant hacked person to not immediately run off for a new one, like making it broken and require actual removal rather than destroying it.

How about call them proxies, because your acting through them.  So you have a controller and proxies.

Mixing real antags, psudo-antags, and crew reacting in odd ways can get fairly toxic.

Add to that the only antag vs antag mode that was ever truly successful was Double agents, and that only really failed because it was anchored to traitor and ops balancing.  So overall this looks like a weird and slightly more flavorful Double-ling

Gangs was a shitshow every round, cult vs cult died in the crib,  Ever Blood vs Clock event learned that they were a terribly boring match up.

Hopeful the fact that the antags are wienie babies and the crew witless pawns will keep the blood levels low, though I expect the same kind of rolling sec assaults and general jackbooted thuggery in regards to cleaning up the webs of victims.

Supermichael777 commented:
As for outing hosts how about instead of giving you a name it gives you an area and time as a victim(so one deconversion wont do it but several will narrowing down the suspects)

and for antag vs antag it gives you a current location once, and outs you specifically if you hit them directly.

Therapist1 commented:
I’m curious, what are your other ideas for powers?

Kierany9 commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,465 @@
+	if(active)
+		user.reset_perspective(null)
+
+/obj/effect/proc_holder/spell/target_hive/hive_see/cast(list/targets, mob/living/user = usr)
```
Probably need to add a sanity check on the spell process to make sure the target isn't qdel'd and to reset vision if it happens. The user can currently do that manually but it's definitely neater for it to work automatically

Kierany9 commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,465 @@
+	charge_max = 600
+	action_icon_state = "force"
+
+/obj/effect/proc_holder/spell/target_hive/hive_force/cast(list/targets, mob/living/user = usr)
```
The control on this is pretty jank and I'm pretty sure targets can resist out of it at full HP while slowed targets are at its mercy

adamkad1 commented:
couldnt you make assimilation mode resembling the borg in star trek?

GuyonBroadway commented:
@adamkad1 as cool as that would be it's kinda not what this pr is about.

Heck assimilators is kinda a bad name for them. They are more like neural ghosts haunting people.

davethwave commented:
The big plan for the hosts is NOT to get discovered until they can fight back against others, so the fact is, is that if they do get discovered then they basically fail. Also for an incentive, what the incentive for a normal human being to care after reporting x antag to sec? There is none besides maybe hoping there's less of a chance of you dying unexpectedly due to greentext needs as even mindshielded people aren't safe anymore once the host grows powerful enough as they have the ability to destroy them.

davethwave commented:
Also, just had a great idea. Can't the tinfoil hat have some kind of ability to resist some of the abilities like assimilate vessel since they are *supposed* to prevent your mind being read and stuff.

Kierany9 commented:
Gamemode's practically ready, only missing some minor (optional) QOL stuff and possibly some polish (nicer sprites, some effects for the high-tier powers and sounds, none of these are my forte though). And another code review.

### Big stuff

* Added Mind Control as a replacement to Cerebellic Pulse, and is unlocked at 15 vessels. Grants you full control over a vessel for 10 seconds. At a hive size of 20/25/30, this increases to 30/60/120 respectively. This should not only add into the thematics of the vessels simply being unwitting pawns in this game of antagonist chess, but makes the vessels themselves far more dangerous, as they could lose control ~~of the situation~~ at any moment.

### Bugfixes

* Fixed merge conflicts.
* Fixed mindshield implants runtiming when used on a hivemind host.
* Network invasion no longer treats you as an enemy hivemind host if a vessel belongs to multiple hives.
* Mobs now require a client to be assimilated

### Lil stuff

* Mind Control and Mass Assimilation are logged and sent to admins.
* Bruteforce now refunds if it gets interrupted at any point.
* Mass assimilation now refunds if it gets interrupted during the first stage.
* Increased number of antags.
* Assimilating an enemy hive host will alert them of the assimilation after a minute and will boot them from the hive after around two minutes.
* Powers now unlock at slightly different hive sizes.
* Added hive size in the antag panel for admins.
* Added a basic tutorial on how the gamemode works when players roll hivemind host.
* Telekinetic Field's walls now last 15 seconds and have a custom name/description.

XDTM commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,571 @@
+					to_chat(user, "<span class='notice'>It seems we have been mistaken, this mind is not the host of a hive.</span>")
+			else
+				to_chat(user, "<span class='userdanger'>Our concentration has been broken, leaving our mind wide open for a counterattack!</span>")
+				to_chat(target, "<span class='userdanger'>Their concentration has been broken... and are wide open for a counterattack!</span>")
```
their concentration are wide open for a counterattack

vuonojenmustaturska commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -0,0 +1,214 @@
+	forge_objectives()
+	..()
+/*
+/datum/antagonist/hivemind/proc/grant_powers()
```
don't leave commented-out code in

Ratvar commented:
Ability to abandon your physical body at 15-30 assimilated/update to Hive Eyes that let's you use abilities from point of view? Become as _meme_, anyone remembers those? 

You already got everyone non-AFK on station, might as well become psionic sentient_virus_2.0
Just swap between hosts and use all abilities from them

- Doesn't look like it will interfere with objectives in any way
- Let's you power tru "you can't change your identity and you have no cool fighting powers so one spearnade..."
- **Endgame superpower**

Or maybe channeled mindswap that also removes random assimilated from network

Like, any slip and those guys are goners: security gang mindshields everyone, mindshielded take all guns and attack all nonmindshielded, normal cult stuff but you are weak AF so you die

What stops them from Repair Protocoling (brainkilling assimilants) from maint locker?

Kierany9 commented:
Four rounds have been run on Bagil so far. While I intended for Hosts to be weak in direct combat, I didn't expect them to end up this generally powerless, nor did I expect the general consensus to be "actually pretty fun" from the antags but "oh shit it wasn't extended?" from the crew. Massive buffs inbound (as) soon (as i can test them locally so they don't introduce any more runtimes):

* Powers unlock far earlier, everything but Medullary Failure, Mass Assimilation and Telekinetic Field is unlocked by 15.
* Made assimilation way easier, it was funny watching people try and get people to stand still but 10 seconds was a bit too much when they aren't expected to go around stunprodding everybody.
* Reworked the way hosts interact with each other. They are no longer warned whenever another host assimilates a vessel of theirs, meaning hosts must proactively search targets with Network Invasion. If it succeeds, the invading host learns the opposing host's identity and the opposing host learns the identity of the just-removed vessel. Assimilating an opposing host will give you their identity once they've booted themselves from the hive, but the message is indistinguishable from the vessel removal message.
* Cooldowns and channelling times have been reduced across the board to incentivize power use.
* Mind Control now partially refunds if it was only used briefly.
* Bruteforce now powers up regular assimilation instead of requiring a minute alone and adjacent to an unmoving target. This should make assimilate objectives somewhat easier as well as allowing for more mind control plays.
* Neural Shock now works like Mind Control, wherein it becomes stronger the larger the hive is, should make Network Invasion way easier to use.
* Adds Distortion Field, an early-game power that causes everybody surrounding a vessel to start hallucinating. Like Mind Control, it gets much stronger with hive size.

Kierany9 commented:
Above changes have been implemented, and some potentially server-crashing runtimes with Mind Control have been fixed.

AnturK commented on code/game/objects/items/implants/implant_mindshield.dm:
```diff
@@ -26,6 +26,16 @@
+				if(hive.hivemembers.Find(target))
+					var/datum/mind/M = hive.owner
+					if(M)
+						warning += "[M.current.real_name]. "
```
Check if the mind still has current at this point.

Kierany9 replied:
Done

AnturK commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -0,0 +1,194 @@
+/datum/antagonist/hivemind/on_gain()
+
+	owner.special_role = special_role
+	apply_innate_effects()
```
This will be done in parent call already, if order is important extract that. But from what i can see it doesn't matter.

Kierany9 replied:
Removed

AnturK commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -0,0 +1,194 @@
+	hive_size = -1
+	check_powers()
+
+	remove_innate_effects()
```
Same thing here.

Kierany9 replied:
Removed

AnturK commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -0,0 +1,194 @@
+	..()
+
+/datum/antagonist/hivemind/proc/add_objective(var/datum/objective/O)
+	objectives += O
```
These wrappers seem bit unnecessary now.

Kierany9 replied:
Removed

AnturK commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -0,0 +1,194 @@
+	if(objectives.len)
+		var/count = 1
+		for(var/datum/objective/objective in objectives)
+			if(objective.check_completion())
```
Use printobjectives instead of writing the whole thing by hand.

Kierany9 replied:
Done

AnturK commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -40,6 +40,11 @@
+	if(is_hivemember(src))
+		remove_hivemember(src)
+	if(is_hivehost(src))
+		var/datum/antagonist/hivemind/hive = src.mind.has_antag_datum(/datum/antagonist/hivemind)
```
No need for src here.

Kierany9 replied:
Removed

AnturK commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -0,0 +1,624 @@
+				return
+		original_body = user
+		vessel = targets[1]
+		backseat = new /mob/living()
```
This could use specific subtype. (Shared by split personality stuff ?)

Kierany9 replied:
The split personality subtype overrides Life() so that if somebody ghosts, the other mob stays forever, and the whole subtype is tied into the owner's trauma. Made a new subtype for it instead.

Kagus commented:
Is there anything planned for an ability to directly implant messages in the mind of an assimilated individual, with the option of disguising it as coming from a particular source? Basically targeted hallucination?

Or would you prefer to not have that kind of direct manipulation of assimilated pawns, in favor of just using them as a kind of passive power source for the real actors, those being the hosts themselves?

davethwave commented:
Just wondering, but how does bruteforce help assimilation?

ATH1909 commented:
You could probably rig up something similar to the hallucination nanite program code for a Hallucination ability, except with more options that would include more/all of the existing hallucinations (make people dodge hallucinatory bullets!).

Maybe you could later make an interaction between this and XDTM's hypnosis PR (only if they both get added, of course)?

ATH1909 commented:
The "become unbodied" thing sounds like a cool idea; maybe add it as a lategame thing. You mentioned mind control being able to do that already; still, it'd be cool to be able to do it permanently (but with no option of going back).

AnturK commented on code/modules/spells/spell_types/hivemind.dm:
```diff
@@ -299,6 +309,7 @@
 			to_chat(user, "<span class='notice'>Our vessel is too far away to control.</span>")
 			revert_cast()
 			return
+		backseat = new /mob/living()
```
Still creating standard living here.

AnturK commented on code/modules/antagonists/hivemind/hivemind.dm:
```diff
@@ -162,31 +154,9 @@
-	else
-		result += "<span class='redtext'>The [special_role_text] has failed!</span>"
-		SEND_SOUND(owner.current, 'sound/ambience/ambifailure.ogg')
+	result += printobjectives(objectives)
```
Print objectives doesn't print the overall win/lose, assuming you still want it.
